src/Features/CSharp/Portable/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,545 @@
+                SwitchExpressionArmSyntax { WhenClause: { } whenClause } switchArm
+                    => CombineWhenClauseCondition(switchArm.Pattern, whenClause.Condition, model),
+                WhenClauseSyntax { Parent: CasePatternSwitchLabelSyntax switchLabel } whenClause
+                    => CombineWhenClauseCondition(switchLabel.Pattern, whenClause.Condition, model),
+                WhenClauseSyntax { Parent: SwitchExpressionArmSyntax switchArm } whenClause
+                    => CombineWhenClauseCondition(switchArm.Pattern, whenClause.Condition, model),
+                _ => null
+            };
+        }
+
+        private static Func<SyntaxNode, SyntaxNode>? CombineLogicalAndOperands(BinaryExpressionSyntax logicalAnd, SemanticModel model)
+        {
+            if (TryDetermineReceiver(logicalAnd.Left, model) is not var (leftReceiver, leftTarget, leftFlipped) ||
+                TryDetermineReceiver(logicalAnd.Right, model) is not var (rightReceiver, rightTarget, rightFlipped))
+            {
+                return null;
+            }
+
+            // If we have an is-expression on the left, first we check if there is a variable designation that's been used on the right-hand-side,
+            // in which case, we'll convert and move the check inside the existing pattern, if possible.
+            // For instance, `e is C c && c.p == 0` is converted to `e is C { p: 0 } c`
+            if (leftTarget.Parent is IsPatternExpressionSyntax isPatternExpression &&
+                TryFindVariableDesignation(isPatternExpression.Pattern, rightReceiver, model) is var (containingPattern, rightNamesOpt))
+            {
+                Debug.Assert(leftTarget == isPatternExpression.Pattern);
+                Debug.Assert(leftReceiver == isPatternExpression.Expression);
+                return root =>
+                {
+                    var rightPattern = CreatePattern(rightReceiver, rightTarget, rightFlipped);
+                    var rewrittenPattern = RewriteContainingPattern(containingPattern, rightPattern, rightNamesOpt);
+                    var replacement = isPatternExpression.ReplaceNode(containingPattern, rewrittenPattern);
+                    return root.ReplaceNode(logicalAnd, AdjustBinaryExpressionOperands(logicalAnd, replacement));
+                };
+            }
+
+            if (TryGetCommonReceiver(leftReceiver, rightReceiver, model) is not var (commonReceiverOpt, leftNames, rightNames))
+                return null;
+
+            return root =>
+            {
+                var leftSubpattern = CreateSubpattern(leftNames, CreatePattern(leftReceiver, leftTarget, leftFlipped));
+                var rightSubpattern = CreateSubpattern(rightNames, CreatePattern(rightReceiver, rightTarget, rightFlipped));
+                // If the common receiver is null, it's an implicit `this` reference in source.
+                // For instance, `prop == 1 && field == 2` would be converted to `this is { prop: 1, field: 2 }`
```

Comment:
this made me curious.  do we have tests that check `base.prop == 1 && base.field == 2`?  we don't want to make that `base is ...`